### PR TITLE
refactor: remove another instance of assuming contiguous shard ids

### PR DESCRIPTION
**Pull request by akhi3030**
See https://github.com/near/nearcore/pull/10230 for prior work.  

CC: @wacban.

### CODE REVIEW

File: core/primitives/src/block.rs [hunk 1]
```diff
@@ -95,22 +93,23 @@ pub enum Block {
...
 
-    (0..num_shards)
-        .map(|i| {
+    shard_ids
+        .into_iter()
+        .map(|&shard_id| {
             let (encoded_chunk, _) = EncodedShardChunk::new(
                 CryptoHash::default(),
-                state_roots[i as usize % state_roots.len()],
+                state_roots[shard_id as usize % state_roots.len()],
```
Comment (nikurt) [inline core/primitives/src/block.rs:109]: I don't understand the `% len` part.
Is it a valid input to have `shard_ids = [1,4,7]`? Then the assertion `state_roots.len() == shard_ids.len()` doesn't help, the shards will be initialized with the same state root.

Comment (akhi3030) [reply core/primitives/src/block.rs:109]: I also do not understand the `% len` part.  I left it as future work to figure out what is happening here. 

> Is it a valid input to have shard_ids = [1,4,7]?

No, in general, we have too many places in the code where we assume that shard ids are contiguous.  We are slowly trying to address all these issues.

Let me know if you any other questions on this PR.

Comment (nikurt) [reply core/primitives/src/block.rs:109]: I don't think that `%len` is correct code. Suggestion below.

```suggestion
   let state_roots : Iter<StateRoot> = if state_roots.len() == shard_ids.len() {
        state_roots.iter()
    } else {
        std::iter::repeat(state_roots[0]).take(4)
    };
    shard_ids.iter().zip(state_roots).map(|(shard_id,state_root){
            let (encoded_chunk, _) = EncodedShardChunk::new(
                CryptoHash::default(),
                state_root,
                CryptoHash::default(),
                genesis_height,
                shard_id,
```

Comment (akhi3030) [reply core/primitives/src/block.rs]: Thanks for the suggestion!  I applied it with some minor edits.

**Review by nikurt — APPROVED** [review]

**Pull request merged**